src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -17,9 +18,62 @@ private Analyzer()
             {
             }
 
-            public static SyntaxKind Analyze(SwitchStatementSyntax node, out bool shouldRemoveNextStatement)
+            public static (SyntaxKind nodeToGenerate, VariableDeclaratorSyntax declaratorToRemoveOpt) Analyze(
+                SwitchStatementSyntax node,
+                SemanticModel semanticModel,
+                out bool shouldRemoveNextStatement)
             {
-                return new Analyzer().AnalyzeSwitchStatement(node, out shouldRemoveNextStatement);
+                var analyzer = new Analyzer();
+                var nodeToGenerate = analyzer.AnalyzeSwitchStatement(node, out shouldRemoveNextStatement);
+
+                if (nodeToGenerate == SyntaxKind.SimpleAssignmentExpression &&
+                    analyzer.TryGetVariableDeclaratorAndSymbol(semanticModel) is var (declarator, symbol))
+                {
+                    if (shouldRemoveNextStatement &&
+                        semanticModel.AnalyzeDataFlow(node.GetNextStatement()).DataFlowsIn.Contains(symbol))
+                    {
+                        // Bail out if data flows into the next statement that we want to move
+                        // For example:
+                        //
+                        //      string name = "";
+                        //      switch (index)
+                        //      {
+                        //          case 0: name = "0"; break;
+                        //          case 1: name = "1"; break;
+                        //      }
+                        //      throw new Exception(name);
+                        //
+                        return default;
+                    }
+
+                    var declaration = declarator.GetAncestor<StatementSyntax>();
+                    if (declaration.Parent == node.Parent && declarator.Initializer is null)
+                    {
+                        var beforeSwitch = node.GetPreviousStatement() is {} previousStatement
```

Comment:
i'd honestly just prefer something like `is StatementSyntax previousStatement`

Reply:
is StatementSyntax doesn't scream null check, this is a usage that's been anticipated for recursive patterns afterall.

Reply:
> is StatementSyntax doesn't scream null check?

It does to me.  Because it just matches `is Whatever` which has been in the language forever and does a null check.  

If we get a `not null` pattern in the future i'd also be ok with that.  I think that `{}` works, but just doesn't read well.  It makes me feel as if it's missing something and trying to say "i need some props...i just forgot what they were :)"

Reply:
>It does to me. Because it just matches is Whatever which has been in the language forever

So does `is object` , perhaps I should quote https://github.com/dotnet/roslyn/pull/37539#discussion_r308862600 from Cyrus

> I agree. I'm actually not a huge fan of is object primarily because, for me, it heavily evokes a type check, when in reality all i'm really trying to convey is the null check in a safe way

I'd agree too.

>If we get a not null pattern in the future i'd also be ok with that

Sadly, `not null` pattern only works when you don't need a variable, you'd need to say 
`not null and var x` to also introduce a variable of a null-checked expression that would err if it's not a reference type - `{}` doesn't.

> I think that {} works, but just doesn't read well. It makes me feel as if it's missing something and trying to say "i need some props...i just forgot what they were :)

There were many requests to have a null-checked var-pattern before C# 8.0 and the answer was to use `{}`, so I'd say that's pretty idiomatic. 

I myself proposed `var x?` similar to swift's "optional patterns" - but it got superseded by `{}` which is a completely accidental feature of recursive patterns.

Reply:
basically, i'd rather wait until we have a good pattern here tbh.

Reply:
I'm with @CyrusNajmabadi and also detest this pattern: it feels like somebody realized you could use a screwdriver to hammer in a nail, and because we've all decided hammers aren't the new hotness we're all trying to use screwdrivers. :-)

Reply:
Note: i'm also very amenable to the idea that i'm being curmudgeonly here :)   I'm not opposed to us migrating to new patterns and practices (lord knows you can see how often i massively do that to some part of the IDE).

That said, it's sometimes good to have a period were we dip our toes in and decide how we feel as a whole and come to some general consensus on a good balance.

i.e. i like patterns.  i like them a lot.  BUt i also think they can become unreadable.  So finding a good balance on how they can be used is great.  But we shouldn't necessarily try to make 100% of all code into a pattern.

This is similar to if we get sequence-expression, and things like return-expressoins.  We'd likely be able to make tons and tons of statements into pure expressions.  But should we?  Probably not.  It will really depend on the code and what the team as a whole feels is a good balance between brevity and readability.

So, in this case, i would use another pattern.  but know that maybe after a few months to a year, the team may start liking this more and more and it may eventually move back :)